### PR TITLE
fix: properly type partial records

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

When we attempt to access an object property with type `Record<T, U>` via dynamic bracket notation the resulting type is `U`, i.e. always defined - TS assumes that if our accessor expression matches type `T` then we get a result of type `U`.

This is not true at runtime. E.g.:

```ts
const opportunityMetadata = stakingOpportunities['notARealId'] // OpportunityMetadata
const opportunityMetadataAssetId = opportunityMetadata.assetId // Throws - attempted to access a property of undefined
```

This PR fixes the types in the Opportunities Slice by using a `PartialRecord` helper (that we should implement more broadly).

Now, continuing our example, when trying to access a property of `opportunityMetadata` the compiler will force us to handle the `undefined` case.

This fixes a whole host of bugs by forcing us to correctly handle `undefined` values.

## Notice

- [X] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Mostly tackles https://github.com/shapeshift/web/issues/3206 (we no longer throw, resulting in a black screen - we never load the modal though).

## Risk

Touches the opportunities slice, though no adverse changes are expected (only less throwing).

## Testing

The opportunities page should be much more robust when values aren't yet (or never will be) loaded.
In particular, refreshing whilst an opportunity overview modal is open should no longer throw, leaving the user with a black screen.

### Engineering

☝️ 

### Operations

☝️ 

## Screenshots (if applicable)

N/A

### CODE REVIEW

File: src/lib/utils.ts [hunk 1]
```diff
@@ -64,3 +64,5 @@ export const isValidAccountNumber = (
   if (accountNumber === null) return false
   return Number.isInteger(accountNumber) && accountNumber >= 0
 }
+
+export type PartialRecord<K extends keyof any, T> = Partial<Record<K, T>>
```
Comment (gomesalexandre) [inline src/lib/utils.ts:68]: Absolute nitpick but what's `T` here ?
I believe we should either use`K, V` generics (Key, value), or the conventional ordered `T, U`

**Review by gomesalexandre — APPROVED** [review]
LGTM, tested the opportunitiesSlice as a sanity check and found no regression

Seems like `PartialRecord` is the gopher to avoid the pitfall of maybe undefined property accesses 🐐 
image

Note, I've updated your description to a `tackles` vs. `closes` WRT https://github.com/shapeshift/web/issues/3206 so we can use this issue to track reloading of opportunities.

**Review by gomesalexandre — APPROVED** [review]
Post-rebase stamp

**Review by gomesalexandre — APPROVED** [review]

File: src/lib/utils.ts (hunk 1 above)
Comment (0xdef1cafe) [reply src/lib/utils.ts:70]: yeah strictly a record is only keyed by `string | number | symbol`

```
> type R = Record<unknown, unknown>
<repl>.ts:4:17 - error TS2344: Type 'unknown' does not satisfy the constraint 'string | number | symbol'.

4 type R = Record<unknown, unknown>
```

Comment (0xdef1cafe) [reply src/lib/utils.ts:70]: nvm `keyof any` *is*` `string | number | symbol` 🤦 

```
➜  web git:(develop) npx ts-node --skipProject
> type K = keyof any
undefined
> const k: K = function() {}
<repl>.ts:5:7 - error TS2322: Type '() => void' is not assignable to type 'string | number | symbol'.

5 const k: K = function() {}
```

Comment (0xdef1cafe) [reply src/lib/utils.ts:70]: but the amount i wish i did this 12 months ago...

Comment (0xApotheosis) [reply src/lib/utils.ts:70]: > but the amount i wish i did this 12 months ago...

We can fix this 😄

Comment (0xApotheosis) [reply src/lib/utils.ts:70]: > Absolute nitpick but what's `T` here ? I believe we should either use`K, V` generics (Key, Value), or the conventional ordered `T, U`

Yeah, you're right there gomes - this should probably be:

```ts
export type PartialRecord<K extends keyof any, V> = Partial<Record<K, V>>
```

**Pull request merged**